### PR TITLE
IWLWIFI Vendor command support

### PR DESCRIPTION
**Pull request by amritara**
Adding Vendor command support in Intel WiFi HAL.

Tracked-On: OAM-79910
Signed-off-by: Jeevaka Prabu Badrappan <jeevaka.badrappan@intel.com>
Signed-off-by: Amrita Raju <amrita.raju@intel.com>

@JeevakaPrabu @sgnanase @swaroopbalan

### CODE REVIEW

File: lib/driver_if.cpp [hunk 1]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+{
+	struct nlattr *cap[MAX_IWL_MVM_VENDOR_ATTR + 1];
+	static struct nla_policy attr_policy[NUM_IWL_MVM_VENDOR_ATTR] = {};
+	char *code = (char *)arg;
+
+	struct nlattr *data = parse_vendor_reply(msg);
+	if (!data)
+		return NL_SKIP;
+
+	attr_policy[IWL_MVM_VENDOR_ATTR_FW_VER].type =	NLA_STRING;
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Remove 1 extra blank space between = and NLA_STRING

File: lib/driver_if.cpp [hunk 2]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+
+	if(nla_parse_nested(cap, MAX_IWL_MVM_VENDOR_ATTR, data, attr_policy)) {
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Introduce a space between if and (

File: lib/driver_if.cpp [hunk 3]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+                hal_printf(MSG_WARNING, "Failed to get fw version");
+                return NL_SKIP;
+        }
+	
+	if (!cap[IWL_MVM_VENDOR_ATTR_FW_VER]) {
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp:1901]: Limit the number of characters/line to 100 characters

File: lib/driver_if.cpp [hunk 4]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+		hal_printf(MSG_DEBUG, "nl80211: No F/w version information available");
+		return NL_SKIP;
+	}
+
+	nla_strlcpy(code, cap[IWL_MVM_VENDOR_ATTR_FW_VER], 50);	//copy attribute to a sized string
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Limit the number of characters/line to 100 characters

File: lib/driver_if.cpp [hunk 5]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+	return NL_SKIP;
+}
+
+
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp:1912]: Remove extra blank line

File: lib/driver_if.cpp [hunk 6]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+	int ret;
+	struct drv_state *drv = (struct drv_state *)handle;
+
+	if (DRV_NOT_INIT(drv, __func__))
+		return -ENODEV;
+
+	buf[0] = '\0';
+
+	ret = drv_vendor_cmd(drv, IWL_MVM_VENDOR_CMD_GET_FW_VERSION,
+	       	NULL, 0, 0, get_fw_handler, buf);
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Can we rename it to get_fw_version_handler?

File: lib/driver_if.cpp [hunk 7]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+        static struct nla_policy attr_policy[NUM_IWL_MVM_VENDOR_ATTR] = {};
+        char *code = (char *)arg;
+
+        struct nlattr *data = parse_vendor_reply(msg);
+        if (!data)
+                return NL_SKIP;
+
+        attr_policy[IWL_MVM_VENDOR_ATTR_DRV_VER].type =  NLA_STRING;
+
+        if(nla_parse_nested(cap, MAX_IWL_MVM_VENDOR_ATTR, data, attr_policy)) {
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Introduce a space between if and (

File: lib/driver_if.cpp [hunk 8]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+                hal_printf(MSG_WARNING, "Failed to get driver version");
+                return NL_SKIP;
+        }
+
+        if (!cap[IWL_MVM_VENDOR_ATTR_DRV_VER]) {
+                hal_printf(MSG_DEBUG, "nl80211: No driver version information available");
+                return NL_SKIP;
+        }
+
+        nla_strlcpy(code, cap[IWL_MVM_VENDOR_ATTR_DRV_VER], 50); //copy attribute to a sized string
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp]: Instead of number can we use MACRO?

File: lib/driver_if.cpp [hunk 9]
```diff
@@ -1842,6 +1871,96 @@ int driver_get_country_code(void *handle, char *code)
...
+	return NL_SKIP;
+
+}
+
+
```
Comment (JeevakaPrabu) [inline lib/driver_if.cpp:1958]: Remove extra blank line

File: lib/wifi_hal.cpp [hunk 10]
```diff
@@ -483,7 +483,14 @@ wifi_error wifi_reset_alert_handler(wifi_request_id id, wifi_interface_handle if
 
 wifi_error wifi_get_firmware_version( wifi_interface_handle iface, char *buffer,
         int buffer_size) {
-    return WIFI_ERROR_NOT_SUPPORTED;
+
+	hal_printf(MSG_DEBUG, "%s", __func__);
+        if ((WIFI_INFO_INVALID(iface, __func__)) || !(buffer && (buffer_size > 0)))
```
Comment (JeevakaPrabu) [inline lib/wifi_hal.cpp]: Indentation doesn't seem to be correct.

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 11]
```diff
@@ -0,0 +1,398 @@
+
+#include "includes.h"
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Copy right information missing

File: wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+#define INTEL_OUI	0x001735
+
+/**
+ * enum iwl_mvm_vendor_cmd - supported vendor commands
+ * @IWL_MVM_VENDOR_CMD_SET_COUNTRY: set a new mcc regulatory information
+ * @IWL_MVM_VENDOR_CMD_RXFILTER: Set/clear rx filter.
+ */
+
+enum iwl_mvm_vendor_cmd {
+	IWL_MVM_VENDOR_CMD_SET_COUNTRY,
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h]: IWL_MVM_VENDOR_CMD_SET_COUNTRY can't be upstreamed. So, this needs to be removed

File: wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+enum iwl_mvm_vendor_rxfilter_op {
+	IWL_MVM_VENDOR_RXFILTER_OP_PASS,
+	IWL_MVM_VENDOR_RXFILTER_OP_DROP,
+};
+
+
+/**
+ * enum iwl_mvm_vendor_attr - attributes used in vendor commands
+ * @__IWL_MVM_VENDOR_ATTR_INVALID: attribute 0 is invalid
+ * @IWL_MVM_VENDOR_ATTR_COUNTRY: MCC to set, for regulatory information (u16)
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+ *      See %iwl_mvm_vendor_rxfilter_flags.
+ * @IWL_MVM_VENDOR_ATTR_RXFILTER_OP: u32 attribute.
+ *      See %iwl_mvm_vendor_rxfilter_op.
+ * @IWL_MVM_VENDOR_ATTR_DRV_VER: string attribute
+ * @IWL_MVM_VENDOR_ATTR_FW_VER: string attribute
+ *
+ */
+enum iwl_mvm_vendor_attr {
+	__IWL_MVM_VENDOR_ATTR_INVALID,
+	IWL_MVM_VENDOR_ATTR_COUNTRY,
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+	IWL_MVM_VENDOR_ATTR_RXFILTER,
+	IWL_MVM_VENDOR_ATTR_RXFILTER_OP,
+	IWL_MVM_VENDOR_ATTR_DRV_VER,
+        IWL_MVM_VENDOR_ATTR_FW_VER,
+
+	NUM_IWL_MVM_VENDOR_ATTR,
+	MAX_IWL_MVM_VENDOR_ATTR = NUM_IWL_MVM_VENDOR_ATTR - 1,
+};
+#define IWL_MVM_VENDOR_FILTER_ARP_NA IWL_MVM_VENDOR_ATTR_FILTER_ARP_NA
+#define IWL_MVM_VENDOR_FILTER_GTK IWL_MVM_VENDOR_ATTR_FILTER_GTK
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h]: Above 2 definitions not needed

File: wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ *****************************************************************************/
+#ifndef __VENDOR_CMD_H__
+#define __VENDOR_CMD_H__
+
+#define INTEL_OUI	0x001735
+
+/**
+ * enum iwl_mvm_vendor_cmd - supported vendor commands
+ * @IWL_MVM_VENDOR_CMD_SET_COUNTRY: set a new mcc regulatory information
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/iwl_vendor_cmd_copy.h]: SET_COUNTRY code can't be up-streamed. So needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+			linkspeed = sig.current_txrate / 1000;
+			ret = os_snprintf(buf, buf_len, "LinkSpeed %d\n", linkspeed);
+		}
+	} 
+	else if (os_strncasecmp(cmd, "RXFILTER-", 9) == 0) {
+		/* Handle all the RXFILTER-*  cmds */
+		ret = wpa_driver_nl80211_rxfilter(priv, cmd + 9);
+	} 
+	else if (os_strncasecmp(cmd, "COUNTRY", 7) == 0) {
+		ret = wpa_driver_nl80211_set_country_code(priv, cmd + 8);
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Remove this and add it to next else loop

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+			ret = -1;
+			goto priv_cmd_end;
+		}
+
+		ret = nl80211_get_link_signal(drv, &sig);
+		if (ret == 0) {
+			rssi = sig.current_signal;
+			ret = os_snprintf(buf, buf_len, "%s rssi %d\n", drv->ssid, rssi);
+		}
+	} else if (os_strcasecmp(cmd, "LINKSPEED") == 0) {
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 19]
```diff
@@ -0,0 +1,398 @@
...
+	u8 macaddr[ETH_ALEN] = {};
+
+	if (os_strcasecmp(cmd, "MACADDR") == 0) {
+		ret = linux_get_ifhwaddr(drv->global->ioctl_sock, bss->ifname, macaddr);
+		if (!ret) {
+			ret = os_snprintf(buf, buf_len,
+					  "Macaddr = " MACSTR "\n", MAC2STR(macaddr));
+		}
+	} else if ((os_strcasecmp(cmd, "RSSI") == 0) ||
+		   (os_strcasecmp(cmd, "RSSI-APPROX") == 0)) {
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 20]
```diff
@@ -0,0 +1,398 @@
...
+		wpa_printf(MSG_DEBUG, "nl80211: vendor command failed err=%d",
+			   ret);
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	return -ENOBUFS;
+}
+
+static int wpa_driver_nl80211_set_country_code(void *priv, char *cmd)
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+					     NL80211_RATE_INFO_BITRATE]) * 100;
+			}
+		}
+	}
+
+	return NL_SKIP;
+}
+
+
+static int nl80211_get_link_signal(struct wpa_driver_nl80211_data *drv,
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 22]
```diff
@@ -0,0 +1,398 @@
...
+			      void *valid_data);
+
+void * nl80211_cmd(struct wpa_driver_nl80211_data *drv,
+			  struct nl_msg *msg, int flags, uint8_t cmd)
+{
+	return (void *)genlmsg_put(msg, 0, 0, drv->global->nl80211_id,
+			   0, flags, cmd, 0);
+}
+
+static int get_link_signal(struct nl_msg *msg, void *arg)
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Needs to be removed

File: wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c [hunk 23]
```diff
@@ -0,0 +1,398 @@
...
+#include "linux_ioctl.h"
+#include "driver_nl80211.h"
+#include "wpa_supplicant_i.h"
+#include "config.h"
+
+#ifdef ANDROID
+#include "android_drv.h"
+#endif
+
+#include "iwl_vendor_cmd_copy.h"
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/driver_cmd_nl80211.c]: Recheck the need for above header files

File: wifi/wifi.c [hunk 24]
```diff
@@ -0,0 +1,746 @@
+/*
```
Comment (JeevakaPrabu) [inline wifi/wifi.c]: This file is not needed at all. Could you remove it?

File: wifi/wpa_supplicant_8_lib/Android.mk [hunk 25]
```diff
@@ -0,0 +1,61 @@
+#
```
Comment (JeevakaPrabu) [inline wifi/wpa_supplicant_8_lib/Android.mk]: Instead of having it under wifi we can directly have it under wpa_supplicant_8_lib

**Review by JeevakaPrabu — CHANGES_REQUESTED** [review]
Commit message needs to be detail the vendor command support added

**Review by JeevakaPrabu — APPROVED** [review]

**Review by swaroopbalan — APPROVED** [review]

**Pull request merged**